test/codeu/chat/server/ConversationAccessServerTest.java

```diff
@@ -0,0 +1,465 @@
+    accessMap = controller.getPermissionMap(conversation);
+    
+    assertFalse(
+        "Check that an owner can ban newcomers",
+        !accessMap.get(newcomer2.id).equals(trollByte));
+    
+    assertFalse(
+         "Check that banned newcomer #2 cannot join conversation",
+         controller.checkMembership(newcomer2.id, conversation));
+
+    controller.addMember(creatorByte, owner.id, conversation);
+    controller.addMember(ownerByte, creator.id, conversation);
+    controller.addMember(ownerByte, member.id, conversation);
+    controller.addMember(memberByte, troll.id, conversation);
+    controller.addMember(trollByte, member.id, conversation);
+    
+    assertFalse(
+        "Check that illegal cases of addMember do nothing",
+        !accessMap.equals(controller.getPermissionMap(conversation)));
+    
+    User newcomer3 = controller.newUser("newcomer3");
+    controller.addMember(creatorByte, newcomer3.id, conversation);
+    accessMap = controller.getPermissionMap(conversation);
+    
+    assertFalse(
+        "Check that a creator can add a newcomer not in chat to their chat",
+        !accessMap.get(newcomer3.id).equals(memberByte));
+    
+    assertFalse(
+         "Check that added newcomer #1 can join conversation",
+         !controller.checkMembership(newcomer3.id, conversation));
+        
+    controller.addMember(creatorByte, badOwner1.id, conversation);
+    accessMap = controller.getPermissionMap(conversation);
+    
+    assertFalse(
+        "Check that a creator can add a banned member back to their chat",
+        !accessMap.get(badOwner1.id).equals(memberByte));
+        
+    assertFalse(
+         "Check that added member #1 can join conversation",
+         !controller.checkMembership(badOwner1.id, conversation));
+        
+    User newcomer4 = controller.newUser("newcomer4");
+    controller.addMember(ownerByte, newcomer4.id, conversation);
+    accessMap = controller.getPermissionMap(conversation);
+    
+    assertFalse(
+        "Check that an owner can add a newcomer not in chat to their chat",
+        !accessMap.get(newcomer4.id).equals(memberByte));
+        
+    assertFalse(
+         "Check that added newcomer #2 can join conversation",
+         !controller.checkMembership(newcomer4.id, conversation));
+    
+    controller.addMember(ownerByte, badOwner2.id, conversation);
+    accessMap = controller.getPermissionMap(conversation);
+    
+    assertFalse(
+        "Check that an owner can add a banned member back to their chat",
+        !accessMap.get(badOwner2.id).equals(memberByte));
+    
+    assertFalse(
+         "Check that added member #2 can join conversation",
+         !controller.checkMembership(badOwner2.id, conversation));
+    
+  }
+  
+  // The following test simulates a more realistic situation where different users have
+  // different roles in the conversations they contribute to.
+  @Test
+  public void multiConversationTest(){
```

Comment:
Same here - you should think about ways to break this method into many smaller tests. I'm happy to suggest some ideas as well.